eng/common/TestResources/New-TestResources.ps1

```diff
@@ -125,374 +125,387 @@ $exitActions = @({
     }
 })
 
-trap {
-    # Like using try..finally in PowerShell, but without keeping track of more braces or tabbing content.
-    $exitActions.Invoke()
+New-Variable -Name 'initialContext' -Value (Get-AzContext) -Option Constant
+if ($initialContext) {
+    $exitActions += {
+        Write-Verbose "Restoring initial context: $($initialContext.Account)"
+        $null = $initialContext | Select-AzContext
+    }
 }
 
-# Enumerate test resources to deploy. Fail if none found.
-$repositoryRoot = "$PSScriptRoot/../../.." | Resolve-Path
-$root = [System.IO.Path]::Combine($repositoryRoot, "sdk", $ServiceDirectory) | Resolve-Path
-$templateFileName = 'test-resources.json'
-$templateFiles = @()
-# Azure SDK Developer Playground
-$defaultSubscription = "faa080af-c1d8-40ad-9cce-e1a450ca5b57"
+# try..finally will also trap Ctrl+C.
+try {
 
-Write-Verbose "Checking for '$templateFileName' files under '$root'"
-Get-ChildItem -Path $root -Filter $templateFileName -Recurse | ForEach-Object {
-    $templateFile = $_.FullName
+    # Enumerate test resources to deploy. Fail if none found.
+    $repositoryRoot = "$PSScriptRoot/../../.." | Resolve-Path
+    $root = [System.IO.Path]::Combine($repositoryRoot, "sdk", $ServiceDirectory) | Resolve-Path
+    $templateFileName = 'test-resources.json'
+    $templateFiles = @()
+    # Azure SDK Developer Playground
+    $defaultSubscription = "faa080af-c1d8-40ad-9cce-e1a450ca5b57"
 
-    Write-Verbose "Found template '$templateFile'"
-    $templateFiles += $templateFile
-}
+    Write-Verbose "Checking for '$templateFileName' files under '$root'"
+    Get-ChildItem -Path $root -Filter $templateFileName -Recurse | ForEach-Object {
+        $templateFile = $_.FullName
 
-if (!$templateFiles) {
-    Write-Warning -Message "No template files found under '$root'"
-    exit
-}
+        Write-Verbose "Found template '$templateFile'"
+        $templateFiles += $templateFile
+    }
 
-$UserName =  if ($env:USER) { $env:USER } else { "${env:USERNAME}" }
-# Remove spaces, etc. that may be in $UserName
-$UserName = $UserName -replace '\W'
+    if (!$templateFiles) {
+        Write-Warning -Message "No template files found under '$root'"
+        exit
+    }
 
-# If no base name is specified use current user name
-if (!$BaseName) {
-    $BaseName = "$UserName$ServiceDirectory"
+    $UserName =  if ($env:USER) { $env:USER } else { "${env:USERNAME}" }
+    # Remove spaces, etc. that may be in $UserName
+    $UserName = $UserName -replace '\W'
+
+    # Make sure $BaseName is set.
+    if ($CI) {
+        $BaseName = 't' + (New-Guid).ToString('n').Substring(0, 16)
```

Comment:
Are we always overriding the parameter here?

Reply:
For CIs, yes. It's been this way for a long time, but was lower down. Figured I'd collocate them for easier maintenance.

Reply:
Daniel and I used to calculate them outside and pass it in, but even back then we realized putting more into the script was easier than maintaining separate scripts (inline or not) in the pipelines.

Reply:
It's a little bit strange. Should we stop passing it all together and only override the value when not provided?

Reply:
No. I do use it for some tests, namely Search because for Search + Key Vault testing it requires that a Key Vault has purge protection, which means I can't purge it and need to use a different name for subsequent testing until it's automatically purged (minimum 7 days, which is what I updated the template to use).

Reply:
Also, for those who want to use this for more long-term resource provisioning (I did that for quite a time and know others who do), I'm sure no one wants to have to remember which t<random digits> resource group is theirs. Maybe we could always tag resources, but it seems like trying to make something that could be easy (username as part of resource names) harder for no other reason than not have a condition in a script (that already has lots of other conditions). I don't see the value in changing this.